Dockerfile

```diff
@@ -13,45 +13,72 @@
-ARG BUILD_DEPS="freetds-dev libkrb5-dev libssl-dev libffi-dev libpq-dev git"
-ARG APT_DEPS="libsasl2-dev freetds-bin build-essential default-libmysqlclient-dev apt-utils curl rsync netcat locales"
+# Airflow extras to be installed
+ARG AIRFLOW_EXTRAS="all"
 
-ENV PATH="$HOME/.npm-packages/bin:$PATH"
+# Increase the value here to force reinstalling Apache Airflow pip dependencies
+ENV FORCE_REINSTALL_ALL_PIP_DEPENDENCIES=1
 
-RUN set -euxo pipefail \
-    && apt update \
-    && if [ -n "${APT_DEPS}" ]; then apt install -y $APT_DEPS; fi \
-    && curl -sL https://deb.nodesource.com/setup_10.x | bash - \
-    && apt update \
-    && apt install -y nodejs \
-    && apt autoremove -yqq --purge \
-    && apt clean
+# Airflow sources change frequently but dependency onfiguration won't change that often
+# We copy setup.py and other files needed to perform setup of dependencies
+# This way cache here will only be invalidated if any of the
+# version/setup configuration change but not when airflow sources change
+COPY setup.py /opt/airflow/setup.py
+COPY setup.cfg /opt/airflow/setup.cfg
 
+COPY airflow/version.py /opt/airflow/airflow/version.py
+COPY airflow/__init__.py /opt/airflow/airflow/__init__.py
+COPY airflow/bin/airflow /opt/airflow/airflow/bin/airflow
+
+# First install only dependencies but no Apache Airflow itself
+# This way regular changes in sources of Airflow will not trigger reinstallation of all dependencies
+# And this Docker layer will be reused between builds.
+RUN pip install --no-cache-dir --upgrade pip==19.0.1 && \
+    pip install --no-cache-dir --no-use-pep517 -e.[$AIRFLOW_EXTRAS]
+
+# Cache for this line will be automatically invalidated if any
+# of airflow sources change
 COPY . /opt/airflow/
 
+# Always add-get update/upgrade here to get latest dependencies before
+# we redo pip install
+RUN apt-get update \
+    && apt-get upgrade -y --no-install-recommends \
+    && apt-get clean && rm -rf /var/lib/apt/lists/*
+
+# PIP Install again including Apache Airflow code now -
+# dependencies should be installed in the previous run
+# We run them just in case some of the apt-get update above
+# causes a conflict
+RUN pip install --no-cache-dir --no-use-pep517 -e .[${AIRFLOW_EXTRAS}]
+
+# Additional python dependencies
+ARG ADDITIONAL_PYTHON_DEPS=""
+
+RUN if [ -n "${ADDITIONAL_PYTHON_DEPS}" ]; then pip install ${ADDITIONAL_PYTHON_DEPS}; fi
+
 WORKDIR /opt/airflow/airflow/www
-RUN npm install \
-    && npm run prod
 
-WORKDIR /opt/airflow
-RUN set -euxo pipefail \
-    && apt update \
-    && if [ -n "${BUILD_DEPS}" ]; then apt install -y $BUILD_DEPS; fi \
-    && if [ -n "${PYTHON_DEPS}" ]; then pip install --no-cache-dir ${PYTHON_DEPS}; fi \
-    && pip install --no-cache-dir --upgrade pip==19.0.1 \
-    && pip install --no-cache-dir --no-use-pep517 -e .[$AIRFLOW_DEPS] \
-    && apt purge --auto-remove -yqq $BUILD_DEPS \
-    && apt autoremove -yqq --purge \
-    && apt clean
+#RUN npm install \
+#    && npm run prod
```

Comment:
I think we're going to need this still - we added it to the CI pipeline recently I think?

Reply:
I disabled this for now during testing, because npm install takes extra time and I do not need it to test some basic behavior. I wil likely uncomment it in final version.